File: packages/core/src/money.test.ts
import { describe, it, expect } from "vitest";
import { toMinorUnits, fromMinorUnits, formatMoney } from "./money";

describe("toMinorUnits", () => {
  it("converts EUR decimals to cents", () => {
    expect(toMinorUnits(0.10, "EUR")).toBe(10);
    expect(toMinorUnits(0.05, "EUR")).toBe(5);
    expect(toMinorUnits(1.00, "EUR")).toBe(100);
    expect(toMinorUnits(9.99, "EUR")).toBe(999);
  });

  it("handles JPY (zero decimal places)", () => {
    expect(toMinorUnits(100, "JPY")).toBe(100);
    expect(toMinorUnits(1, "JPY")).toBe(1);
  });

  it("handles KWD (three decimal places)", () => {
    expect(toMinorUnits(1.000, "KWD")).toBe(1000);
    expect(toMinorUnits(0.100, "KWD")).toBe(100);
  });
});

describe("fromMinorUnits", () => {
  it("converts EUR cents to decimal", () => {
    expect(fromMinorUnits(10, "EUR")).toBeCloseTo(0.10);
    expect(fromMinorUnits(100, "EUR")).toBeCloseTo(1.00);
    expect(fromMinorUnits(999, "EUR")).toBeCloseTo(9.99);
  });
});

describe("round-trip", () => {
  it("toMinorUnits → fromMinorUnits preserves value", () => {
    for (const a of [0.01, 0.05, 0.10, 0.25, 0.50, 1.00, 9.99, 100.00]) {
      expect(fromMinorUnits(toMinorUnits(a, "EUR"), "EUR")).toBeCloseTo(a);
    }
  });
});

describe("formatMoney", () => {
  it("formats EUR values", () => {
    const result = formatMoney({ amount: 10, currency: "EUR" });
    expect(result).toContain("0.10");
  });
});

File: sdk-go/money_test.go
package openpay

import (
	"testing"
)

func TestToMinorUnits(t *testing.T) {
	cases := []struct {
		decimal  float64
		currency string
		want     int64
	}{
		{0.10, "EUR", 10},
		{9.99, "EUR", 999},
		{1.00, "USD", 100},
		{100, "JPY", 100}, // zero decimal
		{1.500, "KWD", 1500}, // three decimals
		{0.001, "KWD", 1},
	}
	for _, c := range cases {
		got := ToMinorUnits(c.decimal, c.currency)
		if got != c.want {
			t.Errorf("ToMinorUnits(%v, %q) = %d, want %d", c.decimal, c.currency, got, c.want)
		}
	}
}

func TestFromMinorUnits(t *testing.T) {
	if got := FromMinorUnits(10, "EUR"); got != 0.10 {
		t.Errorf("FromMinorUnits(10, EUR) = %v, want 0.10", got)
	}
	if got := FromMinorUnits(100, "JPY"); got != 100 {
		t.Errorf("FromMinorUnits(100, JPY) = %v, want 100", got)
	}
}

func TestFormatMoney(t *testing.T) {
	cases := []struct {
		m    Money
		want string
	}{
		{Money{10, "EUR"}, "EUR 0.10"},
		{Money{100, "JPY"}, "JPY 100"},
		{Money{1500, "KWD"}, "KWD 1.500"},
	}
	for _, c := range cases {
		if got := FormatMoney(c.m); got != c.want {
			t.Errorf("FormatMoney(%+v) = %q, want %q", c.m, got, c.want)
		}
	}
}
